src/inmanta/server/services/fileservice.py

```diff
@@ -42,131 +42,86 @@ def __init__(self) -> None:
         if hash_file(content) != file_hash:
             raise BadRequest("The hash does not match the content")
 
-        with open(file_name, "wb+") as fd:
-            fd.write(content)
+        try:
+            await File(content_hash=file_hash, content=content).insert()
+        except UniqueViolationError:
+            raise ServerError("A file with this id already exists.")
 
     @handle(methods.stat_file, file_hash="id")
     async def stat_file(self, file_hash: str) -> Apireturn:
-        file_name = os.path.join(self.server_slice._server_storage["files"], file_hash)
-
-        if os.path.exists(file_name):
+        if await File.has_file_with_hash(file_hash):
             return 200
         else:
             return 404
 
     @handle(methods.get_file, file_hash="id")
     async def get_file(self, file_hash: str) -> Apireturn:
-        content = self.get_file_internal(file_hash)
+        content = await self.get_file_internal(file_hash)
         return 200, {"content": base64.b64encode(content).decode("ascii")}
 
-    def get_file_internal(self, file_hash: str) -> bytes:
+    async def get_file_internal(self, file_hash: str) -> bytes:
         """get_file, but on return code 200, content is not encoded"""
-
-        file_name = os.path.join(self.server_slice._server_storage["files"], file_hash)
-
-        if not os.path.exists(file_name):
+        file: Optional[File] = await File.get_one(content_hash=file_hash)
+        if not file:
             raise NotFound()
-
-        with open(file_name, "rb") as fd:
-            content = fd.read()
-            actualhash = hash_file(content)
-            if actualhash == file_hash:
-                return content
-
-            # handle corrupt file
-            if opt.server_delete_currupt_files.get():
-                LOGGER.error(
-                    "File corrupt, expected hash %s but found %s at %s, Deleting file", file_hash, actualhash, file_name
-                )
-                try:
-                    os.remove(file_name)
-                except OSError:
-                    LOGGER.exception("Failed to delete file %s", file_name)
-                    raise ServerError(
-                        f"File corrupt, expected hash {file_hash} but found {actualhash}. Failed to delete file, please "
-                        "contact the server administrator"
-                    )
-
-                raise ServerError(
-                    f"File corrupt, expected hash {file_hash} but found {actualhash}. "
-                    "Deleting file, please re-upload the corrupt file."
-                )
-            else:
-                LOGGER.error("File corrupt, expected hash %s but found %s at %s", file_hash, actualhash, file_name)
-                raise ServerError(
-                    f"File corrupt, expected hash {file_hash} but found {actualhash}, please contact the server administrator"
-                )
+        return file.content
```

Comment:
When a file is written, we always check whether the hash and the content matches. Since we don't migrate any files on upgrade and the user has to do a full compile to populate the files again, there is no need for a hash check on the read path.

Reply:
Indeed, that's what I meant with "the normal operation flow". But the same goes for the previous normal operation flow and yet we had these checks here. But overall I agree that it serves little purpose.

Reply:
The difference with the previous normal flow is that we added the hash check after we know that corrupt files were present on the filesystem. That's why the check on the read path was required. The current implementation doesn't allow that anymore.